<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Data preprocessing &mdash; alphastats 07-07-2022 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Figures" href="plots.html" />
    <link rel="prev" title="Data import" href="import_data.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> alphastats
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Introduction.html">Alphastats</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="import_data.html">Data import</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Data preprocessing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#removal-of-contaminations">Removal of contaminations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#normalization">Normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#imputation">Imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#subset-data">Subset Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#remove-samples">Remove Samples</a></li>
<li class="toctree-l2"><a class="reference internal" href="#literature">Literature</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="plots.html">Figures</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow_mq.html">Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference/index.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">alphastats</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Data preprocessing</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/data_preprocessing.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="data-preprocessing">
<h1>Data preprocessing<a class="headerlink" href="#data-preprocessing" title="Permalink to this heading"></a></h1>
<p>Preprocessing of the data is substantial for the results of the downstream analysis.
Data preprocessing available in AlphaStats compromises: removal of contaminations, normalization, imputation and subsetting and removal of samples.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">DataSet</span><span class="o">.</span><span class="n">preprocess</span><span class="p">(</span>
        <span class="n">remove_contaminations</span><span class="p">,</span>
        <span class="n">subset</span><span class="p">,</span>
        <span class="n">normalization</span><span class="p">,</span>
        <span class="n">imputation</span><span class="p">,</span>
        <span class="n">remove_samples</span>
<span class="p">)</span>
</pre></div>
</div>
<section id="removal-of-contaminations">
<h2>Removal of contaminations<a class="headerlink" href="#removal-of-contaminations" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>Should I remove contaminations/ <code class="docutils literal notranslate"><span class="pre">remove_contaminations=True</span></code>? Generally speaking - Yes (source)</p></li>
</ul>
<p>Various proteomics software annotate contaminations differently or not at all. MaxQuant describes contaminations or ProteinGroups that should be removed in <code class="docutils literal notranslate"><span class="pre">Only</span> <span class="pre">identified</span> <span class="pre">by</span> <span class="pre">site</span></code>, <code class="docutils literal notranslate"><span class="pre">Reverse</span></code>, <code class="docutils literal notranslate"><span class="pre">Potential</span> <span class="pre">contaminant</span></code>. Likewise MaxQuant, AlphaPept flags spurious proteins as <code class="docutils literal notranslate"><span class="pre">Reverse</span></code>.</p>
<p>In addition, AlphaStats identifies contaminations based on the contamination library created by <a class="reference external" href="https://www.biorxiv.org/content/10.1101/2022.04.27.489766v2.full">Frankenfield et al. 2022</a>.</p>
</section>
<section id="normalization">
<h2>Normalization<a class="headerlink" href="#normalization" title="Permalink to this heading"></a></h2>
<p>Depending on the software and the settings, data could already have been normalized before loading into AlphaStats.</p>
<p>AlphaStats has the following Normalization methods implemented:</p>
<ul class="simple">
<li><p><strong>Z-Score Normalization (Standardization)</strong>: Centers the protein intensity of each sample, meaning scaling the variance to 1.</p></li>
<li><p><strong>Quantile Normalization</strong>: Aims to correct technical bias by adjusting the distribution of protein intensities for each sample. This normalization method is suitable when it is assumed that only a small portion of the protein expression varies among certain conditions, while the majority of the proteome remains stable (<a class="reference external" href="https://doi.org/10.1016/j.biosystems.2022.104661">Dubois et al., 2022</a> ).</p></li>
<li><p><strong>Linear Normalization</strong></p></li>
</ul>
<blockquote>
<div><p><strong>Note</strong>
It has been shown that normalizing the data first and than imputing the data performs better, then the other way around
(<a class="reference external" href="https://doi.org/10.1186/1471-2105-13-S16-S5">Karpievitch et al. 2012</a>). This preprocessing order is also acquired in
AlphaStats (unless preprocessing is done in several steps).</p>
</div></blockquote>
</section>
<section id="imputation">
<h2>Imputation<a class="headerlink" href="#imputation" title="Permalink to this heading"></a></h2>
<p>Especially, missing values are challenging when it comes to analyzing proteomic mass spectrometry data. Missing values can either be missing completely at random (MCAR), due to technical limitations or missing not at random (MNAR) meaning that the abundance is below the detection limit of the platform or completely absent.</p>
<p>To deal with missing values, AlphaStats has the following methods implemented:</p>
<ul class="simple">
<li><p><strong>k-nearest neighbors (kNN)</strong>: Missing values are imputed using the mean protein intensity of k-nearest neighbors</p></li>
<li><p><strong>Random Forest</strong>: Applies the machine learning algorithm random forest and predicts the values of the target variable using specific known target variables as the outcome and other variables as predictors.</p></li>
<li><p><strong>median</strong>: Replaces missing values using the median of each protein.</p></li>
<li><p><strong>mean</strong>: Replaces missing values using the mean of each protein.</p></li>
</ul>
<p>Overall, random forest-based imputation for mass spectrometry data has shown a high performance among several studies compared to other imputations methods(<a class="reference external" href="https://doi.org/10.1186/s12859-019-3110-0">Kokla et al. 2019</a>, <a class="reference external" href="https://doi.org/10.1038/s41598-021-81279-4">Jin et al. 2021</a>). However, when applying random forest imputation to your dataset, you have to expect a long run time.</p>
</section>
<section id="subset-data">
<h2>Subset Data<a class="headerlink" href="#subset-data" title="Permalink to this heading"></a></h2>
<p>In case the proteomics data contains more samples than the metadata, the proteomics data can be filtered based on the samples present in the metadata using <code class="docutils literal notranslate"><span class="pre">DataSet.preprocess(subset=True)</span></code>.</p>
</section>
<section id="remove-samples">
<h2>Remove Samples<a class="headerlink" href="#remove-samples" title="Permalink to this heading"></a></h2>
<p>If you want to remove samples from your <code class="docutils literal notranslate"><span class="pre">DataSet</span></code>, outliers for instance you can give a list of sample names <code class="docutils literal notranslate"><span class="pre">DataSet.preprocess(remove_samples=[&quot;sample1&quot;,</span> <span class="pre">&quot;sample3&quot;])</span></code>.</p>
</section>
<section id="literature">
<h2>Literature<a class="headerlink" href="#literature" title="Permalink to this heading"></a></h2>
<p>Dubois, E., Galindo, A. N., Dayon, L., &amp; Cominetti, O. (2022). Assessing normalization methods in mass spectrometry-based proteome profiling of clinical samples. Bio Systems, 215-216, 104661. https://doi.org/10.1016/j.biosystems.2022.104661</p>
<p>Karpievitch, Y. V., Dabney, A. R., &amp; Smith, R. D. (2012). Normalization and missing value imputation for label-free LC-MS analysis. BMC bioinformatics, 13 Suppl 16(Suppl 16), S5. https://doi.org/10.1186/1471-2105-13-S16-S5</p>
<p>Kokla, M., Virtanen, J., Kolehmainen, M. et al. Random forest-based imputation outperforms other methods for imputing LC-MS metabolomics data: a comparative study. BMC Bioinformatics 20, 492 (2019). https://doi.org/10.1186/s12859-019-3110-0</p>
<p>Jin, L., Bi, Y., Hu, C., Qu, J., Shen, S., Wang, X., &amp; Tian, Y. (2021). A comparative study of evaluating missing value imputation methods in label-free proteomics. Scientific reports, 11(1), 1760. https://doi.org/10.1038/s41598-021-81279-4</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="import_data.html" class="btn btn-neutral float-left" title="Data import" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="plots.html" class="btn btn-neutral float-right" title="Figures" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, MannLabs.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>