<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Workflow Example &mdash; AlphaStats 0.0.5 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API reference for alphastats" href="api_reference/index.html" />
    <link rel="prev" title="Functions" href="functions.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> AlphaStats
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Introduction.html">Alphastats</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="import_data.html">Data import</a></li>
<li class="toctree-l1"><a class="reference internal" href="data_preprocessing.html">Data preprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="functions.html">Functions</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Workflow Example</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#MaxQuant">MaxQuant</a></li>
<li class="toctree-l2"><a class="reference internal" href="#0.-Import-AlphaStats">0. Import AlphaStats</a></li>
<li class="toctree-l2"><a class="reference internal" href="#1.-Import-Data">1. Import Data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#2.-Create-a-DataSet">2. Create a DataSet</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Unnormalized-data,-Sample-Distribution">Unnormalized data, Sample Distribution</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#4.-Visualization">4. Visualization</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Principal-Component-Analysis-(PCA)">Principal Component Analysis (PCA)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api_reference/index.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">AlphaStats</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Workflow Example</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/workflow_mq.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="Workflow-Example">
<h1>Workflow Example<a class="headerlink" href="#Workflow-Example" title="Permalink to this heading"></a></h1>
<section id="MaxQuant">
<h2>MaxQuant<a class="headerlink" href="#MaxQuant" title="Permalink to this heading"></a></h2>
<p>For the analysis we require the <code class="docutils literal notranslate"><span class="pre">proteinGroups.txt</span></code> generated by MaxQuant and a corresponding metadata file where the sample names match with the sample names in the proteinGroups.txt file. At first take a look at the files:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">UserWarning</span><span class="p">,</span> <span class="n">module</span><span class="o">=</span><span class="s2">&quot;openpyxl&quot;</span><span class="p">)</span> <span class="c1"># remove warning from mac</span>
</pre></div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">proteinGroups.txt</span></code> contains all standard column headers from MaxQuant. Later, for our analysis we will use the Protein Intensity described in <code class="docutils literal notranslate"><span class="pre">&quot;LFQ</span> <span class="pre">intensity</span> <span class="pre">[sample]&quot;</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">protein_groups</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;../testfiles/maxquant/proteinGroups.txt&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">low_memory</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">protein_groups</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Protein IDs</th>
      <th>Majority protein IDs</th>
      <th>Peptide counts (all)</th>
      <th>Peptide counts (razor+unique)</th>
      <th>Peptide counts (unique)</th>
      <th>Protein names</th>
      <th>Gene names</th>
      <th>Fasta headers</th>
      <th>Number of proteins</th>
      <th>Peptides</th>
      <th>...</th>
      <th>Potential contaminant</th>
      <th>id</th>
      <th>Peptide IDs</th>
      <th>Peptide is razor</th>
      <th>Mod. peptide IDs</th>
      <th>Evidence IDs</th>
      <th>MS/MS IDs</th>
      <th>Best MS/MS</th>
      <th>Oxidation (M) site IDs</th>
      <th>Oxidation (M) site positions</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>P01911;Q29830;Q9MXZ4;Q3LTJ8;Q3LTJ4;Q3LRY0;Q8HW...</td>
      <td>P01911;Q29830;Q9MXZ4;Q3LTJ8;Q3LTJ4;Q3LRY0;Q8HW...</td>
      <td>4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;...</td>
      <td>4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;4;...</td>
      <td>0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;...</td>
      <td>HLA class II histocompatibility antigen, DRB1-...</td>
      <td>HLA-DRB1;HLA-DR15;HLA-DRB1*;HLA-DRB1*1327;MHC ...</td>
      <td>;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;...</td>
      <td>1834.0</td>
      <td>4.0</td>
      <td>...</td>
      <td>NaN</td>
      <td>0.0</td>
      <td>1287;1288;3971;15222</td>
      <td>True;True;True;True</td>
      <td>1415;1416;4387;16993</td>
      <td>117340;117341;117342;117343;117344;117345;1173...</td>
      <td>56384;56385;56386;159948;159949;159950;602426;...</td>
      <td>56384;56385;159949;602427</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>P05121;A0A024QYT5;B7ZAB0;B7Z4X6;B7Z1D9</td>
      <td>P05121;A0A024QYT5;B7ZAB0;B7Z4X6;B7Z1D9</td>
      <td>10;10;9;8;5</td>
      <td>10;10;9;8;5</td>
      <td>10;10;9;8;5</td>
      <td>Plasminogen activator inhibitor 1</td>
      <td>SERPINE1</td>
      <td>;;;;</td>
      <td>5.0</td>
      <td>10.0</td>
      <td>...</td>
      <td>NaN</td>
      <td>1.0</td>
      <td>1592;3771;4396;4628;6470;7404;9193;11222;13191...</td>
      <td>True;True;True;True;True;True;True;True;True;True</td>
      <td>1746;4166;4871;5128;7155;8188;10148;10149;1257...</td>
      <td>136637;136638;136639;136640;136641;318672;3186...</td>
      <td>64917;149749;149750;178775;178776;184126;25946...</td>
      <td>64917;149749;178775;184126;259465;290355;36587...</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>P55083;A0A024QZ34;K7ES70</td>
      <td>P55083;A0A024QZ34;K7ES70</td>
      <td>2;2;2</td>
      <td>2;2;2</td>
      <td>2;2;2</td>
      <td>Microfibril-associated glycoprotein 4</td>
      <td>MFAP4</td>
      <td>;;</td>
      <td>3.0</td>
      <td>2.0</td>
      <td>...</td>
      <td>NaN</td>
      <td>2.0</td>
      <td>140;15067</td>
      <td>True;True</td>
      <td>155;156;16823</td>
      <td>13174;13175;13176;13177;13178;13179;13180;1318...</td>
      <td>7599;7600;7601;7602;7603;7604;7605;7606;598026...</td>
      <td>7604;598026</td>
      <td>0</td>
      <td>117</td>
    </tr>
    <tr>
      <th>3</th>
      <td>P09972;A0A024QZ64;A8MVZ9;B7Z3K9;B7Z1N6;B7Z3K7;...</td>
      <td>P09972;A0A024QZ64;A8MVZ9;B7Z3K9;B7Z1N6;B7Z3K7;...</td>
      <td>16;16;15;15;13;12;10;10;9;8;7;7;7;5</td>
      <td>13;13;13;12;10;10;9;9;8;5;7;7;6;4</td>
      <td>2;2;2;2;0;2;2;2;2;0;2;2;0;0</td>
      <td>Fructose-bisphosphate aldolase C;Fructose-bisp...</td>
      <td>ALDOC</td>
      <td>;;;;;;;;;</td>
      <td>14.0</td>
      <td>16.0</td>
      <td>...</td>
      <td>NaN</td>
      <td>3.0</td>
      <td>312;749;1545;1675;2085;3265;5046;5708;8847;884...</td>
      <td>True;True;True;False;True;True;True;True;True;...</td>
      <td>346;827;1698;1838;2304;3606;5586;5587;6314;976...</td>
      <td>29006;29007;29008;69091;69092;69093;69094;6909...</td>
      <td>15520;15521;34013;34014;34015;34016;34017;6325...</td>
      <td>15520;34015;63252;67942;84108;131214;198759;21...</td>
      <td>1;2</td>
      <td>40;251</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Q96C19;H0Y4Y4;A0A024QZ77</td>
      <td>Q96C19;H0Y4Y4;A0A024QZ77</td>
      <td>1;1;1</td>
      <td>1;1;1</td>
      <td>1;1;1</td>
      <td>EF-hand domain-containing protein D2</td>
      <td>EFHD2</td>
      <td>;;</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>...</td>
      <td>NaN</td>
      <td>4.0</td>
      <td>8772</td>
      <td>True</td>
      <td>9685</td>
      <td>734937;734938;734939;734940;734941;734942;734943</td>
      <td>342134;342135</td>
      <td>342134</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 2530 columns</p>
</div></div>
</div>
<p>Further, we created a excel-file with the corresponding metadata to our proteinGroups.txt-file. The sample names in the column “sample” match the names in <code class="docutils literal notranslate"><span class="pre">proteinGroups.txt</span></code> file.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s2">&quot;../testfiles/maxquant/metadata.xlsx&quot;</span><span class="p">)</span>
<span class="n">metadata</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>subject external_id</th>
      <th>biological_sample external_id</th>
      <th>sample</th>
      <th>tissue id</th>
      <th>disease id</th>
      <th>intervention id</th>
      <th>tissue</th>
      <th>disease</th>
      <th>biological_sample quantity</th>
      <th>biological_sample quantity_units</th>
      <th>...</th>
      <th>Alanine aminotransferase measurement (34608000)</th>
      <th>Aspartate aminotransferase measurement (45896001)</th>
      <th>Alkaline phosphatase measurement (88810008)</th>
      <th>Gamma glutamyl transferase measurement (69480007)</th>
      <th>Hemoglobin A1c measurement (43396009)</th>
      <th>Total cholesterol:HDL ratio measurement (166842003)</th>
      <th>High density lipoprotein measurement (17888004)</th>
      <th>Low density lipoprotein cholesterol measurement (113079009)</th>
      <th>VLDL cholesterol measurement (104585005)</th>
      <th>Triglycerides measurement (14740000)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>31</td>
      <td>31</td>
      <td>1_31_C6</td>
      <td>BTO:0000131</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>blood plasma</td>
      <td>healthy</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>24.0</td>
      <td>30</td>
      <td>54</td>
      <td>21.0</td>
      <td>6.3</td>
      <td>3.6</td>
      <td>1.26</td>
      <td>2.1</td>
      <td>0.3</td>
      <td>0.58</td>
    </tr>
    <tr>
      <th>1</th>
      <td>32</td>
      <td>32</td>
      <td>1_32_C7</td>
      <td>BTO:0000131</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>blood plasma</td>
      <td>healthy</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>27.0</td>
      <td>28</td>
      <td>27</td>
      <td>38.0</td>
      <td>5.8</td>
      <td>6.6</td>
      <td>1.70</td>
      <td>4.3</td>
      <td>0.6</td>
      <td>1.24</td>
    </tr>
    <tr>
      <th>2</th>
      <td>33</td>
      <td>33</td>
      <td>1_33_C8</td>
      <td>BTO:0000131</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>blood plasma</td>
      <td>healthy</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>18.0</td>
      <td>21</td>
      <td>69</td>
      <td>18.0</td>
      <td>6.2</td>
      <td>5.7</td>
      <td>1.12</td>
      <td>4.1</td>
      <td>0.5</td>
      <td>1.12</td>
    </tr>
    <tr>
      <th>3</th>
      <td>34</td>
      <td>34</td>
      <td>1_34_C9</td>
      <td>BTO:0000131</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>blood plasma</td>
      <td>healthy</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>22.0</td>
      <td>26</td>
      <td>101</td>
      <td>20.0</td>
      <td>6.2</td>
      <td>6.7</td>
      <td>0.91</td>
      <td>4.8</td>
      <td>1.0</td>
      <td>2.20</td>
    </tr>
    <tr>
      <th>4</th>
      <td>35</td>
      <td>35</td>
      <td>1_35_C10</td>
      <td>BTO:0000131</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>blood plasma</td>
      <td>healthy</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>18.0</td>
      <td>25</td>
      <td>61</td>
      <td>13.0</td>
      <td>5.4</td>
      <td>5.5</td>
      <td>1.21</td>
      <td>3.9</td>
      <td>0.4</td>
      <td>0.90</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 46 columns</p>
</div></div>
</div>
</section>
<section id="0.-Import-AlphaStats">
<h2>0. Import AlphaStats<a class="headerlink" href="#0.-Import-AlphaStats" title="Permalink to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">alphastats</span>
</pre></div>
</div>
</div>
</section>
<section id="1.-Import-Data">
<h2>1. Import Data<a class="headerlink" href="#1.-Import-Data" title="Permalink to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">maxquant_data</span> <span class="o">=</span> <span class="n">alphastats</span><span class="o">.</span><span class="n">MaxQuantLoader</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">&quot;../testfiles/maxquant/proteinGroups.txt&quot;</span><span class="p">,</span>
                                          <span class="n">intensity_column</span><span class="o">=</span><span class="s2">&quot;LFQ intensity [sample]&quot;</span><span class="p">,</span>
                                          <span class="n">index_column</span><span class="o">=</span><span class="s2">&quot;Protein IDs&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="2.-Create-a-DataSet">
<h2>2. Create a DataSet<a class="headerlink" href="#2.-Create-a-DataSet" title="Permalink to this heading"></a></h2>
<p>Combine the imported MaxQuant data with the metadata</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span> <span class="o">=</span> <span class="n">alphastats</span><span class="o">.</span><span class="n">DataSet</span><span class="p">(</span>
    <span class="n">loader</span> <span class="o">=</span> <span class="n">maxquant_data</span><span class="p">,</span>
    <span class="n">metadata_path</span> <span class="o">=</span> <span class="s2">&quot;../testfiles/maxquant/metadata.xlsx&quot;</span><span class="p">,</span>
    <span class="n">sample_column</span> <span class="o">=</span> <span class="s2">&quot;sample&quot;</span> <span class="c1"># specify the column that corresponds to the sample names in proteinGroups</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
DataSet has been created.
Attributes of the DataSet can be accessed using:
DataSet.rawdata:         Raw Protein data.
DataSet.mat:    Processed data matrix with ProteinIDs/ProteinGroups as columns and samples as rows. All computations are performed on this matrix.
DataSet.metadata:       Metadata for the samples in the matrix. Metadata will be matched with DataSet.mat when needed (for instance Volcano Plot).
</pre></div></div>
</div>
<p>AlphaStats will create a matrix of the Protein Intensities, which will be accessable using <code class="docutils literal notranslate"><span class="pre">ds.mat</span></code> and will save the metadta as a dataframe <code class="docutils literal notranslate"><span class="pre">ds.metadata</span></code>. Our original MaxQuant ProteinGroup file contains much more samples, than we have metadata for</p>
<p>## 3. Preprocess</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of samples in the matrix: </span><span class="si">{</span><span class="n">ds</span><span class="o">.</span><span class="n">mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">, number of samples in metadata: </span><span class="si">{</span><span class="n">ds</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of samples in the matrix: 312, number of samples in metadata: 48.
</pre></div></div>
</div>
<p>Firstly, we will subset the matrix it will only contains samples, that are also described in the metadata.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="o">.</span><span class="n">preprocess</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of samples in the matrix: </span><span class="si">{</span><span class="n">ds</span><span class="o">.</span><span class="n">mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">, number of samples in metadata: </span><span class="si">{</span><span class="n">ds</span><span class="o">.</span><span class="n">metadata</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of samples in the matrix: 48, number of samples in metadata: 48.
</pre></div></div>
</div>
<section id="Unnormalized-data,-Sample-Distribution">
<h3>Unnormalized data, Sample Distribution<a class="headerlink" href="#Unnormalized-data,-Sample-Distribution" title="Permalink to this heading"></a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">plot_sampledistribution</span><span class="p">(</span><span class="n">color</span> <span class="o">=</span> <span class="s2">&quot;disease&quot;</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">renderer</span> <span class="o">=</span> <span class="s2">&quot;svg&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/workflow_mq_19_0.svg" src="_images/workflow_mq_19_0.svg" /></div>
</div>
<ul class="simple">
<li><p>Contaminations get removed indicated in following columns Only identified by site, Reverse, Potential contaminant (MaxQuant specific) and contamination_library (added by AlphaStats)</p></li>
<li><p>Normalized using quantile normalization</p></li>
<li><p>Missing Values get imputed using K-nearest neighbour imputation</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="o">.</span><span class="n">preprocess</span><span class="p">(</span>
    <span class="n">remove_contaminations</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">normalization</span> <span class="o">=</span> <span class="s2">&quot;quantile&quot;</span><span class="p">,</span>
    <span class="n">imputation</span> <span class="o">=</span> <span class="s2">&quot;knn&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/drq441/opt/anaconda3/lib/python3.9/site-packages/sklearn/preprocessing/_data.py:2590: UserWarning:

n_quantiles (1000) is greater than the total number of samples (48). n_quantiles is set to n_samples.

</pre></div></div>
</div>
<p>The preprocessing steps can be accessed using:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ds</span><span class="o">.</span><span class="n">preprocess_print_info</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
                                                   0                   1
0                  Raw data number of Protein Groups                2611
1                          Matrix: Number of samples                  48
2         Matrix: Number of ProteinIDs/ProteinGroups                2409
3                   Contaminations have been removed                True
4                                      Normalization            quantile
5                                         Imputation  k-Nearest Neighbor
6  Number of removed ProteinGroups due to contami...                 202
</pre></div></div>
</div>
</section>
</section>
<section id="4.-Visualization">
<h2>4. Visualization<a class="headerlink" href="#4.-Visualization" title="Permalink to this heading"></a></h2>
<section id="Principal-Component-Analysis-(PCA)">
<h3>Principal Component Analysis (PCA)<a class="headerlink" href="#Principal-Component-Analysis-(PCA)" title="Permalink to this heading"></a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot</span> <span class="o">=</span> <span class="n">ds</span><span class="o">.</span><span class="n">plot_pca</span><span class="p">(</span><span class="n">group</span> <span class="o">=</span> <span class="s2">&quot;disease&quot;</span><span class="p">,</span> <span class="n">circle</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="n">plot</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">renderer</span> <span class="o">=</span> <span class="s2">&quot;svg&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/workflow_mq_25_0.svg" src="_images/workflow_mq_25_0.svg" /></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="functions.html" class="btn btn-neutral float-left" title="Functions" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api_reference/index.html" class="btn btn-neutral float-right" title="API reference for alphastats" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Elena Krismer.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>